packages/core/src/application.ts

```diff
@@ -3,555 +3,54 @@
-  Binding,
-  Context,
-  Constructor,
-  Provider,
-  inject,
-} from '@loopback/context';
-import {
-  OpenApiSpec,
-  Route,
-  ParsedRequest,
-  OperationObject,
-  ControllerRoute,
-  RouteEntry,
-  createEmptyApiSpec,
-  parseOperationArgs,
-} from '.';
-import {ServerRequest, ServerResponse, createServer} from 'http';
-import {Component, mountComponent} from './component';
-import {getControllerSpec} from './router/metadata';
-import {HttpHandler} from './http-handler';
-import {writeResultToResponse} from './writer';
-import {DefaultSequence, SequenceHandler, SequenceFunction} from './sequence';
-import {RejectProvider} from './router/providers/reject';
-import {
-  FindRoute,
-  InvokeMethod,
-  Send,
-  Reject,
-  ParseParams,
-} from './internal-types';
-import {ControllerClass} from './router/routing-table';
-import {GetFromContextProvider} from './router/providers/get-from-context';
-import {BindElementProvider} from './router/providers/bind-element';
-import {InvokeMethodProvider} from './router/providers/invoke-method';
-import {FindRouteProvider} from './router/providers/find-route';
-import {CoreBindings} from './keys';
-
-const SequenceActions = CoreBindings.SequenceActions;
-
-// NOTE(bajtos) we cannot use `import * as cloneDeep from 'lodash/cloneDeep'
-// because it produces the following TypeScript error:
-//  Module '"(...)/node_modules/@types/lodash/cloneDeep/index"' resolves to
-//  a non-module entity and cannot be imported using this construct.
-const cloneDeep: <T>(value: T) => T = require('lodash/cloneDeep');
-
-const debug = require('debug')('loopback:core:application');
-
-interface OpenApiSpecOptions {
-  version?: string;
-  format?: string;
-}
-
-const OPENAPI_SPEC_MAPPING: { [key: string] : OpenApiSpecOptions; } = {
-  '/openapi.json': {version: '3.0.0', format: 'json'},
-  '/openapi.yaml': {version: '3.0.0', format: 'yaml'},
-  '/swagger.json': {version: '2.0', format: 'json'},
-  '/swagger.yaml': {version: '2.0', format: 'yaml'},
-};
+import { Context, Binding, BindingScope } from '@loopback/context';
+import { Server } from './server';
 
 export class Application extends Context {
-  /**
-   * Handle incoming HTTP(S) request by invoking the corresponding
-   * Controller method via the configured Sequence.
-   *
-   * @example
-   *
-   * ```ts
-   * const app = new Application();
-   * // setup controllers, etc.
-   *
-   * const server = http.createServer(app.handleHttp);
-   * server.listen(3000);
-   * ```
-   *
-   * @param req The request.
-   * @param res The response.
-   */
-  public handleHttp: (req: ServerRequest, res: ServerResponse) => void;
-
-  protected _httpHandler: HttpHandler;
-  protected get httpHandler(): HttpHandler {
-    this._setupHandlerIfNeeded();
-    return this._httpHandler;
-  }
-
-  constructor(public options?: ApplicationOptions) {
+  // Keep a local reference here to simplify iteration over registered
+  // servers at app start.
+  public servers: Map<string, Server>;
```

Comment:
I don't like this approach. So far, methods like `app.controller` were only sugar API for calling `app.bind()` directly, people were able to bind controllers without `app.controller` API. I think we should preserve this functionality for servers too.

If you want a simpler way for iterating over servers, then write a helper method leveraging context iteration under the hood.

This is the use case that I have in mind that will not work in your proposal:

```ts
app.bind('servers.custom').toClass(MyCustomServer);
app.start();
```

Or perhaps:

```ts
class MyComponent {
  constructor() {
    this.providers = {
      'servers.CustomServer': CustomServerProvider,
    }
  }
}

app.component(MyComponent);
app.start();
```